### ISSUE TITLE
LoRA et QLoRA : le fine-tuning abordable

### ISSUE THREAD

**Issue by khalilbenaz** [khalilbenaz/khalilbenaz.github.io#469]
---
lede: LoRA gèle les poids du modèle et n'entraîne que de petites matrices de rang faible — QLoRA y ajoute la quantization 4-bit pour fine-tuner sur un seul GPU grand public.
lede_en: LoRA freezes the base weights and trains only small low-rank matrices — QLoRA adds 4-bit quantization so you can fine-tune a serious model on a single consumer GPU.
title_en: LoRA and QLoRA: affordable fine-tuning
---

## Le problème : le full fine-tuning ne passe pas à l'échelle

Fine-tuner un modèle de 7B paramètres « à l'ancienne » — mettre à jour tous les poids — demande de garder en mémoire les poids, les gradients, et les états de l'optimiseur (Adam en stocke deux par paramètre). En pratique on tourne autour de **16 à 20 octets par paramètre** en mixed precision, soit plus de 100 Go de VRAM pour un 7B. C'est hors de portée d'une seule carte, et déraisonnable en coût cloud pour la plupart des équipes.

Le constat qui a lancé LoRA : quand on adapte un grand modèle à une tâche précise, la **mise à jour** des poids a un rang intrinsèque faible. On n'a pas besoin de bouger tout l'espace des paramètres — quelques directions suffisent.

## Comment marche LoRA

Pour une matrice de poids gelée `W` de dimension `d × k`, LoRA apprend un delta factorisé `ΔW = B · A`, où `A` est `r × k` et `B` est `d × r`, avec `r` (le rang) très petit — typiquement **8, 16 ou 32**. À l'inférence, on calcule :

```
h = W·x + (B·A)·x · (α / r)
```

`W` ne bouge jamais. Seuls `A` et `B` sont entraînés. Pour un 7B avec `r=16` sur les projections d'attention, on entraîne quelques **millions** de paramètres au lieu de milliards — souvent moins de 1 % du total.

Les conséquences pratiques :

- **Empreinte mémoire réduite** — pas d'états d'optimiseur sur les poids gelés.
- **Adaptateurs portables** — un adaptateur LoRA pèse quelques dizaines de Mo. On peut en stocker des dizaines et les échanger à chaud sur le même modèle de base.
- **Pas de dette d'inférence** si on fusionne `B·A` dans `W` après entraînement — latence identique au modèle original.

## QLoRA : la quantization en plus

QLoRA (Dettmers et al., 2023) répond à une question restante : même gelés, les poids de base occupent de la VRAM. La réponse est de **quantizer le modèle de base en 4-bit** (format NF4, NormalFloat 4-bit) et de n'entraîner les adaptateurs LoRA qu'en 16-bit par-dessus.

Trois ingrédients clés :

- **NF4** — un type de données 4-bit conçu pour des poids distribués normalement, plus fidèle que l'int4 naïf.
- **Double quantization** — on quantize aussi les constantes de quantization, grappillant quelques bits par paramètre.
- **Paged optimizers** — on décharge les états de l'optimiseur vers la RAM CPU lors des pics mémoire, évitant les OOM.

Résultat concret : fine-tuner un modèle de la classe 7B–13B **sur un seul GPU 24 Go** (une RTX 3090/4090), là où il fallait un cluster.

## Le compromis honnête

QLoRA n'est pas gratuit. La quantization 4-bit introduit une perte de précision, et l'entraînement à travers des poids quantizés est **plus lent** par étape que le 16-bit pur — on échange du débit contre de la VRAM. Sur des tâches nettes (classification, adaptation de style, format de sortie), la dégradation est souvent négligeable. Sur du raisonnement long ou des tâches très éloignées de la distribution d'origine, un LoRA 16-bit — voire un full fine-tune — peut rester supérieur.

Autres pièges de terrain :

- **Le choix de `r` et `α`** compte moins qu'on ne le craint, mais `α/r` trop élevé déstabilise l'entraînement.
- **Cibler les bonnes couches** : les projections `q`, `k`, `v`, `o` plus les couches MLP donnent en général plus que l'attention seule.
- **Fusionner un adaptateur QLoRA** dans un modèle de base pleine précision demande de déquantizer d'abord — sinon on fige la perte 4-bit.

## Verdict

En 2026, LoRA/QLoRA est le **défaut** pour l'adaptation de modèles ouverts. La question n'est plus « puis-je me le permettre » mais « en ai-je besoin » : souvent le prompt engineering ou le RAG suffisent. Réservez le fine-tuning aux cas où vous devez **encoder un comportement** (format, ton, domaine) que le contexte ne transmet pas de façon fiable — et commencez par QLoRA avant d'escalader.

## The problem: full fine-tuning doesn't scale

Fine-tuning a 7B model the old way — updating every weight — means holding the weights, the gradients, and the optimizer state (Adam keeps two per parameter) in memory at once. In practice that's roughly **16–20 bytes per parameter** in mixed precision, i.e. north of 100 GB of VRAM for a 7B. That's out of reach for a single card and unreasonable in cloud spend for most teams.

The insight behind LoRA: when you adapt a large model to a narrow task, the **update** to the weights has low intrinsic rank. You don't need to move the whole parameter space — a few directions are enough.

## How LoRA works

For a frozen weight matrix `W` of shape `d × k`, LoRA learns a factored delta `ΔW = B · A`, where `A` is `r × k` and `B` is `d × r`, with `r` (the rank) very small — typically **8, 16 or 32**. At inference:

```
h = W·x + (B·A)·x · (α / r)
```

`W` never moves. Only `A` and `B` train. For a 7B with `r=16` on the attention projections, you train a few **million** parameters instead of billions — often under 1% of the total.

What this buys you:

- **Small memory footprint** — no optimizer state on the frozen weights.
- **Portable adapters** — a LoRA adapter is tens of MB. Store dozens and hot-swap them on the same base.
- **No inference tax** if you merge `B·A` into `W` after training — latency identical to the original.

## QLoRA: quantization on top

QLoRA (Dettmers et al., 2023) answers the leftover problem: even frozen, the base weights occupy VRAM. The fix is to **quantize the base model to 4-bit** (NF4, NormalFloat 4-bit) and train the LoRA adapters in 16-bit on top.

Three key ingredients:

- **NF4** — a 4-bit datatype built for normally-distributed weights, more faithful than naive int4.
- **Double quantization** — quantize the quantization constants too, shaving a fraction of a bit per parameter.
- **Paged optimizers** — offload optimizer state to CPU RAM during memory spikes, avoiding OOM.

The concrete payoff: fine-tuning a 7B–13B-class model **on a single 24 GB GPU** (an RTX 3090/4090), where a cluster was needed before.

## The honest trade-off

QLoRA isn't free. 4-bit quantization introduces precision loss, and training through quantized weights is **slower per step** than pure 16-bit — you trade throughput for VRAM. On crisp tasks (classification, style adaptation, output formatting) the degradation is usually negligible. On long-horizon reasoning or tasks far from the base distribution, a 16-bit LoRA — or even a full fine-tune — can still win.

Other field pitfalls:

- **The choice of `r` and `α`** matters less than people fear, but too high an `α/r` destabilizes training.
- **Target the right layers**: the `q`, `k`, `v`, `o` projections plus the MLP layers generally beat attention alone.
- **Merging a QLoRA adapter** into a full-precision base requires dequantizing first — otherwise you bake in the 4-bit loss.

## Verdict

In 2026, LoRA/QLoRA is the **default** for adapting open models. The question is no longer "can I afford it" but "do I need it": prompt engineering or RAG often suffices. Reserve fine-tuning for cases where you must **encode behavior** (format, tone, domain) that context can't reliably convey — and start with QLoRA before escalating.
